### PR TITLE
Add support for "go run" style execution

### PR DESCRIPTION
**Pull request by EvanBoyle**
Addresses https://github.com/pulumi/pulumi/issues/3486

We first look for a named binary, and fall back to "go run" execution if none is found. Passes the `cwd` to the run command, respecting the `-C/--cwd` flag if specified to pulumi.

### CODE REVIEW

File: sdk/go/pulumi-language-go/main.go [hunk 1]
```diff
@@ -92,6 +92,8 @@ func (host *goLanguageHost) GetRequiredPlugins(ctx context.Context,
 	return &pulumirpc.GetRequiredPluginsResponse{}, nil
 }
 
+var unableToFindProgramTemplate = "unable to find program: %s"
```
Comment (pgavlin) [inline sdk/go/pulumi-language-go/main.go]:
```suggestion
const unableToFindProgramTemplate = "unable to find program: %s"
```

File: sdk/go/pulumi-language-go/main.go [hunk 2]
```diff
@@ -135,18 +137,44 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
...
 
+	// by default we try to run a named executable on the path, but we will fallback to 'go run' style execution
+	goRunInvoke := false
+
 	// The program to execute is simply the name of the project.  This ensures good Go toolability, whereby
 	// you can simply run `go install .` to build a Pulumi program prior to running it, among other benefits.
 	program, err := findProgram(req.GetProject())
 	if err != nil {
-		return nil, errors.Wrap(err, "problem executing program (could not run language executor)")
+		message := "problem executing program (could not run language executor)"
```
Comment (pgavlin) [inline sdk/go/pulumi-language-go/main.go]:
```suggestion
		const message = "problem executing program (could not run language executor)"
```

File: sdk/go/pulumi-language-go/main.go [hunk 3]
```diff
@@ -135,18 +137,44 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
...
 
 	logging.V(5).Infof("language host launching process: %s", program)
 
 	// Now simply spawn a process to execute the requested program, wiring up stdout/stderr directly.
 	var errResult string
-	cmd := exec.Command(program)
+	var cmd *exec.Cmd
+
+	if goRunInvoke {
+		cwd, err := os.Getwd()
```
Comment (pgavlin) [inline sdk/go/pulumi-language-go/main.go]: I think you also need to take into account the requested working directory for the project, if any: https://github.com/pulumi/pulumi/blob/master/sdk/proto/language.proto#L47

You can fetch this via `req.GetPwd()`. I believe that it is a path that is relative to the current working directory (or the empty string if unset).

Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go]: Empirically, this works as is. I've added an integration test that uses the `main` field in `Pulumi.yaml` to verify this behavior.

File: sdk/go/pulumi-language-go/main.go [hunk 4]
```diff
@@ -135,18 +137,50 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
...
 
+	// by default we try to run a named executable on the path, but we will fallback to 'go run' style execution
+	goRunInvoke := false
+
 	// The program to execute is simply the name of the project.  This ensures good Go toolability, whereby
 	// you can simply run `go install .` to build a Pulumi program prior to running it, among other benefits.
 	program, err := findProgram(req.GetProject())
 	if err != nil {
-		return nil, errors.Wrap(err, "problem executing program (could not run language executor)")
+		const message = "problem executing program (could not run language executor)"
```
Comment (lukehoban) [inline sdk/go/pulumi-language-go/main.go]: Just `message :=`.

File: sdk/go/pulumi-language-go/main.go [hunk 5]
```diff
@@ -135,18 +137,50 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
 		return nil, errors.Wrap(err, "failed to prepare environment")
 	}
 
+	// by default we try to run a named executable on the path, but we will fallback to 'go run' style execution
+	goRunInvoke := false
+
 	// The program to execute is simply the name of the project.  This ensures good Go toolability, whereby
 	// you can simply run `go install .` to build a Pulumi program prior to running it, among other benefits.
```
Comment (lukehoban) [inline sdk/go/pulumi-language-go/main.go]: This comment probably warrants updating.

File: sdk/go/pulumi-language-go/main.go [hunk 6]
```diff
@@ -135,18 +137,50 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
...
+		if err.Error() == fmt.Sprintf(unableToFindProgramTemplate, req.GetProject()) {
+			logging.V(5).Infof("Unable to find program %s in $PATH, attempting invocation via 'go run'", program)
+			program, err = findProgram("go")
+			if err != nil {
+				return nil, errors.Wrap(err, message)
+			}
+			goRunInvoke = true
+		} else {
+			return nil, errors.Wrap(err, message)
+		}
```
Comment (lukehoban) [inline sdk/go/pulumi-language-go/main.go]: Is this same technique something we can do for .NET as well to accomplish the opposite?

Would love to have a consistent approach here if possible.

File: pkg/testing/integration/program.go [hunk 7]
```diff
@@ -1608,6 +1611,12 @@ func (pt *programTester) prepareGoProject(projinfo *engine.Projinfo) error {
 	if err != nil {
 		return err
 	}
+
+	// skip building if the 'go run' invocation path is requested.
+	if pt.opts.GoRunInvoke {
```
Comment (lukehoban) [inline pkg/testing/integration/program.go]: We have a `RunBuild` flag already - make sense to reuse that here?

File: sdk/go/pulumi-language-go/main.go [hunk 8]
```diff
@@ -135,18 +137,50 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
...
-	cmd := exec.Command(program)
+	var cmd *exec.Cmd
+
+	if goRunInvoke {
+		cwd, err := os.Getwd()
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to get current working directory")
+		}
+
+		goFileSearchPattern := fmt.Sprintf("%s/*.go", cwd)
```
Comment (lukehoban) [inline sdk/go/pulumi-language-go/main.go]: Does filepath.Glob use Unix style paths?  Or do we need filepath.Join here?

File: sdk/go/pulumi-language-go/main.go (hunk 4 above)
Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go]: You and @pgavlin are [in opposition here](https://github.com/pulumi/pulumi/pull/3503#discussion_r345994737)

File: sdk/go/pulumi-language-go/main.go (hunk 8 above)
Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go]: Good catch.

File: sdk/go/pulumi-language-go/main.go (hunk 6 above)
Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go]: Great idea, can you take a look at https://github.com/pulumi/pulumi/issues/3509  and verify that I've captured your thoughts accurately. 

Happy to pull this onto my plate if we think it's an immediate priority.

File: sdk/go/pulumi-language-go/main.go (hunk 5 above)
Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go]: Done

File: pkg/testing/integration/program.go (hunk 7 above)
Comment (EvanBoyle) [reply pkg/testing/integration/program.go]: I missed that, thanks for pointing it out. With this change, the default behavior will be for go programs to run without an explicit build step. I've updated the existing go integration test cases to maintain their current behavior, and to ensure we continue to have coverage on that path.

**Pull request merged**